src/Microsoft.DotNet.XHarness.Android/AdbRunner.cs

```diff
@@ -193,26 +195,23 @@ public int InstallApk(string apkPath)
 
             var result = RunAdbCommand($"install \"{apkPath}\"");
 
-            // If this keeps happening, we should look into exercising Helix infra retry logic when on Helix
-            // since users should be able assume tests themselves can't break the ADB server.
-            if (result.ExitCode == (int)AdbExitCodes.ADB_BROKEN_PIPE)
+            // Two possible retry scenarios, theoretically both can happen on the same run:
+
+            // 1. Pipe between ADB server and emulator device is broken; restarting the ADB server helps
+            if (result.ExitCode == (int)AdbExitCodes.ADB_BROKEN_PIPE || result.StandardError.Contains(AdbInstallBrokenPipeError))
             {
                 _log.LogWarning($"Hit broken pipe error; Will make one attempt to restart ADB server, then retry the install");
                 KillAdbServer();
                 StartAdbServer();
-                needToRetry = true; 
+                result = RunAdbCommand($"install \"{apkPath}\"");
             }
 
+            // 2. Installation cache on device is messed up; restrting the device reliably seems to unblock this (unless the device is actually full, if so this will error the same)

```

Comment:
```suggestion
            // 2. Installation cache on device is messed up; restarting the device reliably seems to unblock this (unless the device is actually full, if so this will error the same)

```